### PR TITLE
Feature/issue 231 phone numbers

### PR DESCRIPTION
**Pull request by bensinjin**

### CODE REVIEW

**Comment by bensinjin** [comment]
@rasmus-storjohann if you can please run an import locally and give me your feedback on some of the numbers that get skipped due to failing validation and how you'd like to handle them.

**Comment by rasmus-storjohann-PG** [comment]
Remove `.DS_Store` files

File: bc211/parser.py [hunk 1]
```diff
@@ -226,3 +229,44 @@ def parse_state_province(address):
...
+        intl_phone_number = convert_bc_phone_number_to_international(phone_number)
+    except ValueError:
+        LOGGER.warning('Failed to parse value: "%s" for phone number.', phone_number)
+        return False
+    try:
+        validate_international_phonenumber(intl_phone_number)
+    except ValidationError:
+        LOGGER.warning('Failed to parse value: "%s" for international phone number.', intl_phone_number)
+        return False
+    return True
```
[the remark this answers is not in the document]
Comment (rasmus-storjohann-PG) [reply bc211/parser.py]: What if we changed the tack for "invalid" phone numbers and instead imported everything (if it's a non-empty string, it's valid) and left it to the client to recognize phone numbers and make them clickable? There are so many cases of strings that are hard to parse automatically but easy to parse for someone with basic English skills... Thoughts?

File: bc211/importer.py [hunk 2]
```diff
@@ -154,3 +166,19 @@ def create_location_address(location, address, address_type):
...
+        phone_number_type, created = PhoneNumberType.objects.get_or_create(
+            id=dto.phone_number_type_id
+        )
+        if created:
+            counters.count_phone_number_types()
+            LOGGER.debug('PhoneNumberType: "%s"', phone_number_type.id)
+        number = PhoneNumber.objects.create(
+            location=location,
+            phone_number_type=phone_number_type,
+            phone_number='+' + str(dto.phone_number)
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply bc211/importer.py]: This is defined by the library we use see https://github.com/stefanfoulis/django-phonenumber-field. By default it expects a string representing an international number. The field itself is a Charfield in the db not an int as previously suggested.

[the remark this answers is not in the document]
Comment (rasmus-storjohann-PG) [reply bc211/importer.py]: That library is probably not able to represent numbers such as "604-946-6622 Local 312" or "1-800-394-3472 or 613-944-1310". So if we go with importing all these various strings into the phone number field, that means not using that library. What benefit does the library get us?

File: bc211/parser.py [hunk 3]
```diff
@@ -101,9 +104,9 @@ def parse_spatial_location_if_defined(site):
         return None
     return dtos.SpatialLocation(latitude=latitude, longitude=longitude)
 
-def parse_services(site, organization_id, site_id):
+def parse_service_list(site, organization_id, site_id):
     services = site.findall('SiteService')
-    return map(ServiceParser(organization_id, site_id), services)
+    return list(map(ServiceParser(organization_id, site_id), services))
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply bc211/parser.py]: I've created a validator for an optional list of objects which can I'm now using on this dto. Its implementation requires a list.

File: bc211/parser.py [hunk 4]
```diff
@@ -226,3 +229,44 @@ def parse_state_province(address):
...
+        return False
+    try:
+        validate_international_phonenumber(intl_phone_number)
+    except ValidationError:
+        LOGGER.warning('Failed to parse value: "%s" for international phone number.', intl_phone_number)
+        return False
+    return True
+
+def convert_phone_type_to_type_id(phone_type):
+    return phone_type.lower().replace(' ', '_')
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply bc211/parser.py]: Yeah I was thinking we shouldn't store separate entries in our db for what we'd expect to be the same thing.

File: bc211/parser.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bensinjin) [reply bc211/parser.py]: In order to run the phone number through our phone number field's validator (this comes from the aforementioned library, it's not our code) we need to first ensure we're working with an integer otherwise validation automatically fails. It's not ideal but it's probably better than writing the international phone validator ourselves.

In terms of not validating. I personally I think we probably need to validate the phone numbers as I don't think we can assume that the text from these fields will be simple enough to not need translation. I'd like to confirm how many are failing validation though just to see what we're dealing with in terms of coercing them to valid values.

File: bc211/validate.py [hunk 5]
```diff
@@ -25,6 +25,17 @@ def optional_object(the_class, field, values):
         return value
     raise exceptions.InvalidNestedObjectXmlParseException(field)
 
+def optional_list_of_objects(the_class, field, values):
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply bc211/validate.py]: I agree. I've put tests for this in another file called test_validate.py as apposed to testing them indirectly through a dto test.

File: human_services/phone_numbers/models.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+from django.db import models
+from phonenumber_field.modelfields import PhoneNumberField
+from common.models import RequiredCharField
+from human_services.locations.models import Location
+
+
+class PhoneNumberType(models.Model):
+    id = RequiredCharField(primary_key=True, max_length=200)
+
+class PhoneNumber(models.Model):
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply human_services/phone_numbers/models.py]: Fixed

File: bc211/dtos.py [hunk 7]
```diff
@@ -53,3 +54,9 @@ def __init__(self, **kwargs):
         self.address_type_id = validate.required_string('address_type_id', kwargs)
         self.state_province = validate.optional_string('state_province', kwargs)
         self.postal_code = validate.optional_string('postal_code', kwargs)
+
+class PhoneNumber:
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply bc211/dtos.py]: Fixed

File: bc211/tests/test_parser.py [hunk 8]
```diff
@@ -351,3 +351,86 @@ def test_address_line_parser_returns_none_when_line_1_is_empty(self):
...
+                </Phone>
+            </Site>'''
+        )
+        phone_numbers = parser.parse_site_phone_number_list(root, site_id)
+        self.assertEqual(len(phone_numbers), 0)
+
+    def test_parses_phone_into_expected_dto_object(self):
+        site_id = a_string()
+        phone_type = a_string()
+        phone_number = an_integer()
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply bc211/tests/test_parser.py]: Fixed

File: bc211/validate.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rasmus-storjohann-PG) [reply bc211/validate.py]: What is the single responsibility of DTO classes. They transmit valid data between producers and consumers of that data. Emphasis on *valid*. It's a good style for DTOs to validate their content at construction, then all downstream code can make strong assumptions about any data that they receive. It follows that the tests should be written against the DTO classes. The fact that these classes use a helper function under the hood is an implementational detail that the tests don't need to be aware of.

File: human_services/phone_numbers/views.py [hunk 9]
```diff
@@ -0,0 +1,3 @@
+from django.shortcuts import render
+
+# Create your views here.
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply human_services/phone_numbers/views.py]: Fixed

File: human_services/phone_numbers/tests/helpers.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        self.id = a_string()
+
+    def with_id(self, type_id):
+        self.id = type_id
+        return self
+
+    def build(self):
+        address_type = PhoneNumberType()
+        address_type.id = self.id
+        return address_type
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply human_services/phone_numbers/tests/helpers.py]: Right, fixed

**Comment by bensinjin** [comment]
Ok @rasmus-storjohann I've made the phone number field a text field in the db and removed validation along with the other things we've discussed. Please take another look.

File: bc211/tests/test_parser.py [hunk 11]
```diff
@@ -351,3 +351,86 @@ def test_address_line_parser_returns_none_when_line_1_is_empty(self):
...
+    def test_does_not_parse_phone_with_alphanumeric_phone_number(self):
+        site_id = a_string()
+        phone_type = a_string()
+        phone_number = a_string()
+        xml = self.build_phone_xml(phone_number, phone_type)
+        root = etree.fromstring(xml)
+        phone_numbers = parser.parse_site_phone_number_list(root, site_id)
+        self.assertEqual(len(phone_numbers), 0)
+
+    def test_convert_bc_phone_to_intl_returns_int(self):
```
[the remark this answers is not in the document]
Comment (bensinjin) [reply bc211/tests/test_parser.py]: Removed as we're no longer doing validation

File: bc211/tests/test_dtos.py [hunk 12]
```diff
@@ -78,3 +138,27 @@ def test_throws_on_missing_taxonomy_id(self):
...
+
+    def test_throws_on_missing_location_id(self):
+        with self.assertRaises(exceptions.MissingRequiredFieldXmlParseException):
+            dtos.PhoneAtLocation(phone_number_type_id=a_string(), phone_number=a_phone_number())
+
+    def test_throws_on_missing_phone_number_type_id(self):
+        with self.assertRaises(exceptions.MissingRequiredFieldXmlParseException):
+            dtos.PhoneAtLocation(phone_number=a_phone_number(), location_id=a_string())
+
+    def test_throws_on_phone_number(self):
```
Comment (rasmus-storjohann-PG) [inline bc211/tests/test_dtos.py]: rename test to indicate why it throws

**Comment by bensinjin** [comment]
Ok @rasmus-storjohann this #230 is in. Please one final review one the most recent commit.
